build/report-datagen.js

```diff
@@ -1,168 +1,219 @@
+ * 2. Geography key (e.g. 37063000101)
+ * 3. Metric name (e.g. "AFTERNOON-COOL-ISLANDS")
+ *
+ * The value contained under the last key is an object structured as follows:
+ *
+ * ```json
+ * {
+ *  "map": {y_2012: value, ...},
+ *  "w": {y_2012: value, ...}
+ * }
+ * ```
+ */
+async function loadAndTransposeAllMetrics(metricGeographyPairs) {
+  const metrics = {};
+
+  await Promise.all(
+    metricGeographyPairs.map(async ({ geographyId, metricConfig: { metric: metricName } }) => {
+      if (!(geographyId in metrics)) {
+        metrics[geographyId] = {};
       }
-    }
-  });
 
-  // /////////////////////////////////////////////
-  // Create report-specific JSON files.
-  // /////////////////////////////////////////////
+      const metricFilePath = path.join(inputBase, `${geographyId}/m${metricName}.json`);
 
-  const allGeographyMetricsCached = {}; // All metric values for current geography.
-  const countyAverages = {}; // All county averages for all metrics.
+      // Read metric file
+      let metricFileContents;
+      try {
+        metricFileContents = await fs.promises.readFile(metricFilePath);
+      } catch (err) {
+        console.error(`Error reading ${metricFilePath}: ${err.message}`);
+        throw err;
+      }
+
+      // Parse metric data
+      let metricData;
+      try {
+        metricData = JSON.parse(metricFileContents);
+      } catch (err) {
+        console.error(`Error parsing ${metricFilePath}: ${err.message}`);
+        throw err;
+      }
 
-  // Create one big array of each unique metric, geography level pair that we can iterate through.
-  const metricGeographyPairs = siteConfig.geographies.flatMap(
-    geography => Object.values(dataConfig)
-      .filter(m => (m.geographies.indexOf(geography.id) > -1))
-      .map(metric => ({ geography, metric })),
+      // Transpose metric for each geography
+      Object.entries(metricData.map).forEach(([geographyKey, yearsData]) => {
+        if (!(geographyKey in metrics[geographyId])) {
+          metrics[geographyId][geographyKey] = {};
+        }
+        metrics[geographyId][geographyKey][metricName] = {
+          map: yearsData,
+        };
+        if ("w" in metricData) {
+          metrics[geographyId][geographyKey][metricName].w = metricData.w[geographyKey];
+        }
+      });
+    })
   );
 
-  async.each(metricGeographyPairs, (pair, callback) => {
-    const { geography, metric } = pair;
-    if (!(geography.id in allGeographyMetricsCached)) {
-      allGeographyMetricsCached[geography.id] = {};
+  return metrics;
+}
+
+/**
+ * Compute averages for each metric across the entire county (all geographies).
+ *
+ * - Uses `world_val` from the data config as the county average for a metric when it is set.
+ * - Otherwise, county averages are computed from metric data. The order of the `geographies` array in
+ *   the site config determines the precedence of which geography level serves as the source for the
+ *   aggregation.
```

Comment:
@timstallmann could you confirm that it is true that the order of the geographies in the site config does/should determine the precedence of the geography levels in this aggregation?

Reply:
Oh good call! That sounds right to me, I hadn't quite ever noticed that behavior.